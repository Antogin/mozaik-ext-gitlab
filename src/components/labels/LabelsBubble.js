import React, { Component } from 'react'
import LabelsIcon from 'react-icons/lib/fa/tags'
import { TrapApiError, Widget, WidgetHeader, WidgetBody, WidgetLoader } from '@mozaik/ui'
import { ResponsiveBubble } from 'nivo'
import { labelsPropTypes, labelsDefaultProps } from './propTypes'
import { countLabel } from './counts'

export default class LabelsBubbles extends Component {
    static propTypes = labelsPropTypes

    static defaultProps = labelsDefaultProps

    static getApiRequest({ project }) {
        return {
            id: `gitlab.projectLabels.${project}`,
            params: { project },
        }
    }

    render() {
        const { apiData, apiError, title, countBy, animate } = this.props

        let body = <WidgetLoader />
        let subject = null
        let count = 0
        if (apiData) {
            const { project, labels } = apiData

            count = labels.length

            const data = {
                name: 'labels',
                color: '#000',
                children: labels,
            }

            subject = (
                <a href={project.web_url} target="_blank">
                    {project.name}
                </a>
            )

            body = (
                <ResponsiveBubble
                    root={data}
                    labelSkipRadius={12}
                    value={countBy}
                    label={d => `${d.name} ${d[countBy]}`}
                    labelTextColor="inherit:darker(1.6)"
                    leavesOnly={true}
                    colorBy={d => d.color}
                    animate={animate}
                />
            )
        }

        return (
            <Widget>
                <WidgetHeader
                    title={title || `Labels by ${countLabel(countBy)}`}
                    subject={title ? null : subject}
                    count={count}
                    icon={LabelsIcon}
                />
                <WidgetBody>
                    <TrapApiError error={apiError}>
                        {body}
                    </TrapApiError>
                </WidgetBody>
            </Widget>
        )
    }
}
