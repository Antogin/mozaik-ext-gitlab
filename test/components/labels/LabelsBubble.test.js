import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { defaultTheme } from '@mozaik/ui'
import LabelsBubble from './../../../src/components/labels/LabelsBubble'
import { sampleProject, sampleLabels } from './sampleData'

test('should render as expected', () => {
    const tree = renderer.create(
        <ThemeProvider theme={defaultTheme}>
            <LabelsBubble
                project={sampleProject.name}
                apiData={{ project: sampleProject, labels: sampleLabels }}
                theme={defaultTheme}
            />
        </ThemeProvider>
    )

    expect(tree).toMatchSnapshot()
})
