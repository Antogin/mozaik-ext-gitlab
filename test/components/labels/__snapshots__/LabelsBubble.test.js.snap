// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
<div
  className="sc-bdVaJa eydJRb"
  style={Object {}}
>
  <div
    className="sc-bwzfXH fMILzD"
  >
    <header
      className="sc-htpNat llvRri"
      style={Object {}}
    >
      <span>
        <div
          className="sc-ifAKCX bDEKRH"
        >
          <a
            href="https://gitlab.com/plouc/ploucifier"
            target="_blank"
          >
            ploucifier
          </a>
        </div>
        Labels by open issues
        <div
          className="sc-bxivhb eDwDKI"
        >
          3
        </div>
      </span>
      <div
        className="sc-EHOje dlEdwN"
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": "#000",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m9.5 10.5q0-1.2-0.8-1.9t-1.9-0.8-1.9 0.8-0.8 1.9 0.8 1.9 1.9 0.8 1.9-0.8 0.8-1.9z m22.7 12.2q0 1.1-0.8 1.9l-10.4 10.5q-0.8 0.8-2 0.8-1.1 0-1.9-0.8l-15.2-15.2q-0.8-0.8-1.3-2.2t-0.6-2.5v-8.8q0-1.1 0.8-1.9t1.9-0.8h8.9q1.1 0 2.5 0.5t2.1 1.4l15.2 15.2q0.8 0.8 0.8 1.9z m8.2 0q0 1.1-0.8 1.9l-10.5 10.5q-0.8 0.8-1.9 0.8-0.8 0-1.2-0.3t-1.2-1l10-10q0.8-0.8 0.8-1.9 0-1.1-0.8-1.9l-15.2-15.2q-0.8-0.8-2.1-1.4t-2.5-0.5h4.7q1.2 0 2.5 0.5t2.2 1.4l15.2 15.2q0.8 0.8 0.8 1.9z"
            />
          </g>
        </svg>
      </div>
    </header>
    <div
      className="sc-bZQynM cUinle"
      style={Object {}}
    >
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <svg
            height={400}
            width={600}
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              transform="translate(10,10)"
            >
              <circle
                className="nivo_bubble_node"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                r={100.58708222343544}
                style={
                  Object {
                    "fill": "#F00",
                    "stroke": "#F00",
                    "strokeWidth": 0,
                  }
                }
                transform="translate(204.73567289408075,166.58466790532816)"
              />
              <circle
                className="nivo_bubble_node"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                r={85.0116005118636}
                style={
                  Object {
                    "fill": "#F0F",
                    "stroke": "#F0F",
                    "strokeWidth": 0,
                  }
                }
                transform="translate(391.3262167851995,166.58466790532816)"
              />
              <circle
                className="nivo_bubble_node"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                r={72.7995853639637}
                style={
                  Object {
                    "fill": "#00F",
                    "stroke": "#00F",
                    "strokeWidth": 0,
                  }
                }
                transform="translate(311.9369644907279,304.11922489614494)"
              />
              <text
                className="nivo_bubble_legend"
                dy={5}
                style={
                  Object {
                    "fill": Rgb {
                      "b": 0,
                      "g": 0,
                      "opacity": 1,
                      "r": 144.11097101734254,
                    },
                    "pointerEvents": "none",
                  }
                }
                textAnchor="middle"
                transform="translate(204.73567289408075,166.58466790532816)"
              >
                label A 21
              </text>
              <text
                className="nivo_bubble_legend"
                dy={5}
                style={
                  Object {
                    "fill": Rgb {
                      "b": 144.11097101734254,
                      "g": 0,
                      "opacity": 1,
                      "r": 144.11097101734254,
                    },
                    "pointerEvents": "none",
                  }
                }
                textAnchor="middle"
                transform="translate(391.3262167851995,166.58466790532816)"
              >
                label B 15
              </text>
              <text
                className="nivo_bubble_legend"
                dy={5}
                style={
                  Object {
                    "fill": Rgb {
                      "b": 144.11097101734254,
                      "g": 0,
                      "opacity": 1,
                      "r": 0,
                    },
                    "pointerEvents": "none",
                  }
                }
                textAnchor="middle"
                transform="translate(311.9369644907279,304.11922489614494)"
              >
                label C 11
              </text>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
